plugins/woocommerce/src/Internal/PushNotifications/DataStores/PushTokensDataStore.php

```diff
@@ -169,25 +149,172 @@ public function update( &$push_token ) {
 	}
 
 	/**
-	 * Returns an array of post meta objects as key => value pairs.
+	 * Deletes a push token.
+	 *
+	 * @since 10.5.0
+	 * @param PushToken $push_token An instance of PushToken.
+	 * @throws InvalidArgumentException If the token can't be deleted.
+	 * @throws PushTokenNotFoundException If the token can't be found.
+	 * @return bool
+	 */
+	public function delete( PushToken &$push_token ): bool {
+		if ( ! $push_token->can_be_deleted() ) {
+			throw new InvalidArgumentException(
+				'Can\'t delete push token because the push token data provided is invalid.'
+			);
+		}
+
+		$post = get_post( $push_token->get_id() );
+
+		if ( ! $post || PushToken::POST_TYPE !== $post->post_type ) {
+			throw new PushTokenNotFoundException( 'Push token could not be found.' );
+		}
+
+		wp_delete_post( (int) $push_token->get_id(), true );
+
+		return true;
+	}
+
+	/**
+	 * Find tokens for this user and platform that match either the token
+	 * or device UUID. We check the token value to avoid creating a duplicate.
+	 * We check the device UUID value because only one token should be issued
+	 * per device, therefore if we already have one then we can update it to
+	 * avoid creating a duplicate.
+	 *
+	 * @since 10.5.0
+	 * @param PushToken $push_token An instance of PushToken.
+	 * @return null|PushToken
+	 * @throws InvalidArgumentException If push token is missing data.
+	 */
+	public function get_by_token_or_device_id( PushToken &$push_token ): ?PushToken {
+		if (
+			! $push_token->get_user_id()
+			|| ! $push_token->get_platform()
+			|| ! $push_token->get_origin()
+			|| (
+				/**
+				 * Platforms iOS and Android require token OR device UUID.
+				 */
+				$push_token->get_platform() !== PushToken::PLATFORM_BROWSER
+				&& ! $push_token->get_token()
+				&& ! $push_token->get_device_uuid()
+			)
+			|| (
+				/**
+				 * Browsers don't have device UUIDs, so require token.
+				 */
+				$push_token->get_platform() === PushToken::PLATFORM_BROWSER
+				&& ! $push_token->get_token()
+			)
+		) {
+			throw new InvalidArgumentException(
+				'Can\'t retrieve push token because the push token data provided is invalid.'
+			);
+		}
+
+		global $wpdb;
+
+		$posts = $wpdb->get_results(
+			$wpdb->prepare(
+				"SELECT *
+				FROM {$wpdb->posts}
+				WHERE post_type = %s
+					AND post_status = 'private'
+					AND post_author = %d
+				ORDER BY ID DESC",
+				PushToken::POST_TYPE,
+				$push_token->get_user_id()
+			)
+		);
```

Comment:
I believe I mentioned this before, but just in case, unless we're doing a special query that will be much faster, I think this should just be a regular `WP_Query`. All the fields here seem query-able using that API.